### PR TITLE
[JENKINS-37566] - Always run setAccessible and classloader constructors in the privileged blocks

### PR DESCRIPTION
**Pull request by oleg-nenashev**
- [x] - Always run setAccessible() from an privileged block (via Util class)
- [x] - Create Classloaders inside privileged blocks

Just fixes about 7 issues reported by FindBugs.

https://issues.jenkins-ci.org/browse/JENKINS-37566

### CODE REVIEW

File: src/main/java/hudson/remoting/Which.java [hunk 1]
```diff
@@ -141,19 +144,25 @@ public static File jarFile(Class clazz) throws IOException {
                 Object delegate = is;
                 while (delegate.getClass().getEnclosingClass()!=ZipFile.class) {
                     Field f = delegate.getClass().getDeclaredField("delegate");
-                    f.setAccessible(true);
+                    if (!ReflectionUtils.makeAccessibleOrLog(f, LOGGER, Level.FINE)) {
+                        continue;
```
Comment (oleg-nenashev) [inline src/main/java/hudson/remoting/Which.java]: Self-:bug: causes infinite cycle

File: src/main/java/hudson/remoting/Which.java [hunk 2]
```diff
@@ -141,19 +144,25 @@ public static File jarFile(Class clazz) throws IOException {
...
+                    }
+
                     delegate = f.get(delegate);
                     //JENKINS-5922 - workaround for CertificateReaderInputStream; JBoss 5.0.0, EAP 5.0 and EAP 5.1
                     if(delegate.getClass().getName().equals("java.util.jar.JarVerifier$VerifierStream")){
                         f = delegate.getClass().getDeclaredField("is");
-                        f.setAccessible(true);
+                        if (!ReflectionUtils.makeAccessibleOrLog(f, LOGGER, Level.FINE)) {
+                            continue;
```
Comment (oleg-nenashev) [inline src/main/java/hudson/remoting/Which.java]: Maybe better to stop execution as well

File: src/main/java/org/jenkinsci/remoting/nio/SelectableFileChannelFactory.java [hunk 3]
```diff
@@ -128,7 +131,7 @@ public SocketChannel create(FileDescriptor fd) {
         } catch (ClassNotFoundException e) {
```
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/remoting/nio/SelectableFileChannelFactory.java]: Multi-catch around should be cleaned up during the refactoring

File: src/test/java/org/jenkinsci/remoting/engine/HandlerLoopbackLoadStress.java [hunk 4]
```diff
@@ -45,14 +45,7 @@
...
 import java.nio.channels.SocketChannel;
-import java.security.KeyManagementException;
-import java.security.KeyPair;
-import java.security.KeyPairGenerator;
-import java.security.KeyStore;
-import java.security.KeyStoreException;
-import java.security.MessageDigest;
-import java.security.NoSuchAlgorithmException;
-import java.security.UnrecoverableKeyException;
+import java.security.*;
```
Comment (oleg-nenashev) [inline src/test/java/org/jenkinsci/remoting/engine/HandlerLoopbackLoadStress.java]: Self-:bug:
Tried editing something in IDEA, going back to my caves

**Comment by oleg-nenashev** [comment]
@reviewbybees since the test failure is unrelated

File: src/main/java/org/jenkinsci/remoting/util/ReflectionUtils.java [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+    }
+
+    /**
+     * Makes a method accessible using reflection.
+     * There is no automatic accessibility status recovery.
+     * @param object Object to be modified
+     * @throws PrivilegedActionException Missing permissions or other modification error.
+     *      This exception cannot happen if the object is accessible when the method is invoked.
+     */
+    public static void makeAccessible(final @Nonnull AccessibleObject object) throws PrivilegedActionException {
```
Comment (kohsuke) [inline src/main/java/org/jenkinsci/remoting/util/ReflectionUtils.java]: :bug:  This is a wrong usage of `AccessController.doPrivileged()` that creates a security hole. It allows untrusted caller to use this public method to elevate privilege and make any method/field accessible.

It's best not to fiddle with `AcessController` IMO. The only situation where it makes a difference is when remoting is in trusted code source and the client app code is not trusted. Remoting deals too much with classloader and other potentially dangerous code. It's best just not to let the untrusted caller use remoting at all.

Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/remoting/util/ReflectionUtils.java]: Trusted caller can always do it via the reflection API, no?

Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/remoting/util/ReflectionUtils.java]: Anyway, restriction of this Utility class access is reasonable

Comment (kohsuke) [inline src/main/java/org/jenkinsci/remoting/util/ReflectionUtils.java]: > Trusted caller can always do it via the reflection API, no?

I'm not sure how to interpret that.

A decade ago, I worked on JAXB, which eventually became a part of JavaSE. A part of that integration work involves making JAXB safely usable from untrusted code (think of applet/JavaFX, for example), and that experience taught me a lot about the danger and difficulty of correctly using privilege escalation.

The findbugs explanation of `DP_DO_INSIDE_DO_PRIVILEGED` does not discuss any implication of privilege escalations and down right dangerous.

IIUC, the motivation here is to make this module FindBugs clean, and in that case I recommend simply disabling this detector.

I can tell you more about this if you really care about, but it's much too long to type it in.

Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/remoting/util/ReflectionUtils.java]: That's why I submitted it as a separate PR. I went through some docs/discussions on this topic and decided it may be reasonable to perform such change. 

The motivation is to really make FindBugs clean, but I agree it requires more investigation. Putting it on hold

**Comment by oleg-nenashev** [comment]
We may hit it in Java9, but I have no plan to work on it anytime soon

File: src/main/java/hudson/remoting/RemoteClassLoader.java [hunk 6]
```diff
@@ -105,14 +109,29 @@
...
             // as ClassLoaderProxy. This means, the so-called remote classloader here is
             // actually our classloader that we exported to the other side.
             return ((ClassLoaderProxy)proxy).cl;
         }
-        return new RemoteClassLoader(parent, proxy);
+
+        // Create new RemoteClassLoader otherwise
+        final RemoteClassLoader created;
+        try {
+            created = AccessController.doPrivileged(
```
Comment (jglick) [inline src/main/java/hudson/remoting/RemoteClassLoader.java:123]: You should only ever need to use this in case you are running with a security manager enabled, which for Remoting we are not. So I do not see the point of this PR. If it is just to get SpotBugs to shut up, use the warning suppression annotation, or just configure a SpotBugs exclude for that warning in the POM.

**Review by jglick — CHANGES_REQUESTED** [review]

File: src/main/java/hudson/remoting/RemoteClassLoader.java (hunk 6 above)
Comment (jglick) [reply src/main/java/hudson/remoting/RemoteClassLoader.java:123]: And there is a Java JEP proposing to deprecate the security manager, and associated methods like this.

**Comment by jtnord** [comment]
Usb tbtime to close this PR given comments from KK and Jesse?

**Comment by jeffret-b** [comment]
@oleg-nenashev , I'm going to close this, based on the various comments and how outdated it is. If you still want to do something with it you can reopen it.
